apps/www/_blog/2023-11-21-oauth2-login-python-flask-apps.mdx

```diff
@@ -0,0 +1,140 @@
+
+## Initiate the client
+
+Create another file called `supabase_client.py` and in this file, we will initiate our Supabase client.
+
+```python
+import os
+from flask import g
+from werkzeug.local import LocalProxy
+from supabase.client import Client, ClientOptions
+from flask_storage import FlaskSessionStorage
+
+url = os.environ.get("SUPABASE_URL", "")
+key = os.environ.get("SUPABASE_KEY", "")
+
+def get_supabase() -> Client:
+    if "supabase" not in g:
+        g.supabase = Client(
+            url,
+            key,
+            options=ClientOptions(
+                storage=FlaskSessionStorage(),
+                flow_type="pkce"
+            ),
+        )
+    return g.supabase
+
+supabase: Client = LocalProxy(get_supabase)
+```
+
+Let's focus on the `get_supabase` function. Here we are checking if we have an instance of the client stored in our global object `g`, if not we create the client and store it in the global object under the `supabase` name. You will notice in the `ClientOptions` that we are specifying the `FlaskSessionStorage` class we created earlier and we are also specifying a very important option that allows us to handle the OAuth flow on the server side, the `flow_type="pkce"`.
+
+## Sign in with GitHub
+
+Supabase Auth supports Sign in with GitHub on the web, native Android applications, and Chrome extensions.
+
+For detailed set up and implementation instructions please refer to the [docs](https://supabase.com/docs/guides/auth/social-login/auth-github).
+
+## Create sign-in route
+
+Inside our application code `app.py`, we can create the sign-in route to trigger the OAuth sign-in request.
+
+```python
+@app.route("/signin/github")
+def signin_with_github():
+    res = supabase.auth.sign_in_with_oauth(
+        {
+            "provider": "github",
+            "options": {
+	            "redirect_to": f"{request.host_url}callback"
+	        },
+        }
+    )
+    return redirect(res.url)
+```
+
+In this function `options` object we specify a `redirect_to` parameter which will point to the callback route we will create in the next step. This function will generate a url for us to use to redirect the user to, in this case we are using `github` as our OAuth provider so we will be redirected to the GitHub OAuth consent screen.
+
+## Create callback route
+
+Let's add another route to our `app.py` file for the callback endpoint we specified in our sign in route.
+
+```python
+@app.route("/callback")
+def callback():
+    code = request.args.get("code")
+    next = request.args.get("next", "/")
+
+    if code:
+        res = supabase.auth.exchange_code_for_session({"auth_code": code})
+
+    return redirect(next)
+```
+
+Here we are getting the `code` query parameter from the request object, if this is available we then exchange the code for a session so that the user will be signed in. Under the hood the `supabase` python library will handle storing this session (JWT) into a cookie and sign the user in.
```

Comment:
```suggestion
Here we're getting the `code` query parameter from the request object, if this is available we then exchange the code for a session so that the user will be signed in. Under the hood the `supabase` python library will handle storing this session (JWT) into a cookie and sign the user in.
```